### ISSUE TITLE
Research user and organisation metadata provisioning 

### ISSUE THREAD

**Issue by janhalen** [OS2sandbox/sandbox-samtale#4]
### Could user provisioning via [System for Cross-domain Identity Management](https://scim.cloud/) be the way to go?

Using Authentik as IDP via OpenID connect to authenticate and authorize, there seems to be support for provisioning via a [SCIM provider](https://docs.goauthentik.io/docs/providers/scim/)

I also found that Zulip has "beta" support for provisioning via SCIM https://zulip.readthedocs.io/en/stable/production/scim.html.

This seems like the most standardized way of bringing over users (and maybe groups or roles?) 

Thoughts?

**Comment by janhalen** [comment]
A bit of further perspectives from the Authentik team: https://goauthentik.io/blog/2023-10-05-scims-many-deviations/

**Comment by mikkeschiren** [comment]
I talked with @colin-campbell about this, and he is not so keen on using SCIM, it has it own issues.

**Comment by janhalen** [comment]
Relates to an older issue from the identity-service issue-tracker #6

**Comment by janhalen** [comment]
> I talked with @colin-campbell about this, and he is not so keen on using SCIM, it has it own issues.

I have also read critisism of SCIM, but mostly how the implementation is done, I dont have further data.

It also seems like the maintainer of the upstream Authentik project (and CTO of the Public Benefit Company, sponsoring the project) agrees with these sentiments, as expressed in this blog: https://goauthentik.io/blog/2023-10-05-SCIMs-many-deviations.

With focus on open standards, reusability and robust maintenance strategies, I am interrested in discussion what our options are for provisioning entities like Zulip channels and user mappings to those channels in a sustainable way.

**Comment by janhalen** [comment]
### A non-standard provisioning metod possibility?
An application specific metod to managing Zulip seems to center around [management commands](https://zulip.readthedocs.io/en/latest/production/management-commands.html)

I found [this entry point](https://chat.zulip.org/#narrow/stream/138-user-questions/topic/.E2.9C.94.20Automating.20organization.20deployment ) on chat.zulip.org by [Tim Abbot](https://github.com/timabbott) describing how this technical concept could be explored as a provisioning solution.

Interresting metods for the provisioning could be: 

- [create a channel](https://github.com/zulip/zulip/blob/main/zerver/management/commands/create_stream.py)
- [add users to channel](https://github.com/zulip/zulip/blob/main/zerver/management/commands/create_stream.py)

Important nomenclature remark: It seems that references to the concept of _"streams"_ in the code are equivalent to  _"channels"_ in the UI and docs.

We do need to be aware of the maintainability tasks this could bring.

 The primary user stories are centered around pretty "realtime-ish" updating of organizational de/reattachments and offboarding scenarios. E.g. see some (not-yet evaluated) user-stories here: 

https://github.com/OS2sandbox/sandbox-myndighedsidentitet-issues/issues/76
https://github.com/OS2sandbox/sandbox-myndighedsidentitet-issues/issues/70

**Comment by janhalen** [comment]
### A modern REST API standard method

Theres also an API possibility, especially https://zulip.com/api/subscribe seems to be a fitting endpoint to use

**Comment by mikkeschiren** [comment]
On the CLI-level (manage.py), I tested out creating the organisation on first startup, in the same way, provisioning could be done of channels (streams) and adding users to those streams.

**Comment by janhalen** [comment]
Currently Anna-Lis is doing everything manually... For the next more "reproducible step" we should try and scope a solution..

Running manage.py inside the container is not a very cloud-native approach though?

A change to the OIDC backend to map OIDC token natively in zulip might take weeks to develop and get a PR in?

Creating a sync container talking to the Authentik API and the Zulip API might be possible, but there is the issue of maintainability..

@mikkeschiren: We should really discuss this and scope it.

@ChatBotBerg: Compared to the manual time you are currently putting in.. how do you feel that this issue should be prioritized...

**Comment by ChatBotBerg** [comment]
@janhalen  & @mikkeschiren this issue needs to be prioritized in some form as i am spending tripple time handling user roles.

### User Story:

_As an IT administrator responsible for user access in Zulip, i experience that all users are assigned the role member in zulip regardless ogf the userroles assigned in authentic._

_I need for Zulip to automatically assign users to the correct groups based on their identity provider claims (e.g., OIDC groups from Authentik),_

_So that I don’t have to manually click through each user profile and assign them to groups every time someone joins, changes department, or leaves the organization._


**MOST important**: that users registred as external users - automatically appears as GUESTS in zulip and not as members.

**_Nice to have:_** that i can add a standard authentic role to users  that provides them with the moderator role on Zulip.

**Comment by janhalen** [comment]
@mikkeschiren: 

I that a task Digitalist could handle? How many hours are needed?

And if it should be "made right" as an upstream contribution to Zulip, should we bring in a specialist supplier or do you have internal Python / Django expertise?

Making the upstream change i guess would start from the official docs: https://zulip.readthedocs.io/en/latest/production/authentication-methods.html

**Comment by mikkeschiren** [comment]
@janhalen - yes we could look into this, I will get into the code for this tomorrow.

**Comment by mikkeschiren** [comment]
I will do an investigation today and on Friday, I will have an estimate on the work on Monday_tuseday next week.

**Comment by mikkeschiren** [comment]
I have done some progress, will keep you in the loop

**Comment by janhalen** [comment]
> I have done some progress, will keep you in the loop

Fantastic! @ChatBotBerg will be pleased to be replaced by a robot army soon 😄

**Comment by janhalen** [comment]
@mikkeschiren: A collaboration effort to align OS2Adgang towards KeyCloak (as Collin originally proposed) is generating steam.. Would that be something that Digitalist would be interrested in collaborating on? 

And in the end, would that create some opportunities for solving this issue?

**Comment by mikkeschiren** [comment]
Yes, that we would be. And sorry for the late answer here, I missed this question, my Github notification box got at bit overloaded... 
And yes, Keycloak would be good fit to try to solve the issue.

**Comment by janhalen** [comment]
@ChatBotBerg: To eliminate the manual registration process, i recommend supporting this issue, and make a plan for how it could be implemented.
